### PR TITLE
sendmsg/recvmsg: Added support for msg_control on Windows

### PR DESCRIPTION
**Pull request by mitza-oci**

### CODE REVIEW

File: ACE/ace/OS_NS_sys_socket.cpp [hunk 1]
```diff
@@ -271,4 +271,85 @@ ACE_OS::send_partial_i (ACE_HANDLE handle,
...
+
+  return 0;
+}
+
+int ACE_OS::sendmsg_win32_i (ACE_HANDLE handle,
+                             msghdr const *msg,
+                             int flags,
+                             unsigned long &bytes_sent)
+{
+#  if _WIN32_WINNT >= 0x0600
```
Comment (jrw972) [inline ACE/ace/OS_NS_sys_socket.cpp]: Why is there no corresponding _WIN32_WINNT check in recvmsg?

File: ACE/ace/SOCK_Dgram.cpp [hunk 2]
```diff
@@ -339,7 +345,7 @@ ACE_SOCK_Dgram::recv (iovec iov[],
 #endif /* ACE_HAS_SOCKADDR_MSG_NAME */
   recv_msg.msg_namelen = addr.get_size ();
 
-#if defined (ACE_HAS_4_4BSD_SENDMSG_RECVMSG)
+#ifdef ACE_USE_MSG_CONTROL
```
Comment (jrw972) [inline ACE/ace/SOCK_Dgram.cpp]: Closing comment is no longer correct.

File: ACE/ace/SOCK_Dgram.cpp [hunk 3]
```diff
@@ -353,24 +359,22 @@ ACE_SOCK_Dgram::recv (iovec iov[],
   addr.set_size (recv_msg.msg_namelen);
   addr.set_type (((sockaddr_in *) addr.get_addr())->sin_family);
 
-#if defined (ACE_HAS_4_4BSD_SENDMSG_RECVMSG)
+#ifdef ACE_USE_MSG_CONTROL
```
Comment (jrw972) [inline ACE/ace/SOCK_Dgram.cpp]: Close comment is no longer correct.

File: ACE/ace/OS_NS_sys_socket.cpp (hunk 1 above)
Comment (mitza-oci) [reply ACE/ace/OS_NS_sys_socket.cpp]: sendmsg is ifdef'ed in the platform headers, recvmsg isn't in platform headers at all (ioctl)

File: ACE/ace/SOCK_Dgram.cpp (hunk 2 above)
Comment (mitza-oci) [reply ACE/ace/SOCK_Dgram.cpp]: Good reason not to use comments for things that the editor/IDE should take care of.

File: ACE/ace/OS_NS_sys_socket.cpp (hunk 1 above)
Comment (mitza-oci) [reply ACE/ace/OS_NS_sys_socket.cpp]: I think we also have the option of ioctl for sendmsg, but not sure if that's how we want to go.

**Comment by jwillemsen** [comment]
@mitza-oci It looks we have some new warnings triggered by this PR, see http://buildlogs.remedy.nl/win_mingw64_64_tao/2019_10_30_08_31_Brief.html#section_5

```
[Details] X:/BUILD/DOC_ROOT/ACE/ace/OS_NS_string.inl:312:20: warning: 'char* strncpy(char*, const char*, size_t)' specified bound 128 equals destination size [-Wstringop-truncation]
[Details] x:/BUILD/DOC_ROOT/ACE/ace/OS_NS_sys_socket.cpp: In function 'void* ACE_6_5_6::ACE_OS::accept(ACE_HANDLE, sockaddr*, int*, const ACE_6_5_6::ACE_Accept_QoS_Params&)':
[Details] x:/BUILD/DOC_ROOT/ACE/ace/OS_NS_sys_socket.cpp:25:90: warning: cast between incompatible function types from 'ACE_6_5_6::ACE_QOS_CONDITION_FUNC' {aka 'int (*)(iovec*, iovec*, ACE_6_5_6::ACE_QoS*, ACE_6_5_6::ACE_QoS*, iovec*, iovec*, unsigned int*, long unsigned int)'} to 'LPCONDITIONPROC' {aka 'int (*)(_WSABUF*, _WSABUF*, _QualityOfService*, _QualityOfService*, _WSABUF*, _WSABUF*, unsigned int*, long long unsigned int)'} [-Wcast-function-type]
[Details] X:/BUILD/DOC_ROOT/ACE/ace/OS_NS_macros.h:28:34: note: in definition of macro 'ACE_SOCKCALL_RETURN'
[Details] x:/BUILD/DOC_ROOT/ACE/ace/SOCK_Dgram.cpp: In member function 'ssize_t ACE_6_5_6::ACE_SOCK_Dgram::recv(iovec*, int, ACE_6_5_6::ACE_Addr&, int, ACE_6_5_6::ACE_INET_Addr*) const':
[Details] X:/BUILD/DOC_ROOT/ACE/ace/os_include/sys/os_socket.h:120:30: warning: comparison of integer expressions of different signedness: 'int' and 'long long unsigned int' [-Wsign-compare]
[Details] x:/BUILD/DOC_ROOT/ACE/ace/SOCK_Dgram.cpp:367:27: note: in expansion of macro 'ACE_CMSG_FIRSTHDR'
[Details] X:/BUILD/DOC_ROOT/ACE/ace/os_include/sys/os_socket.h:120:30: warning: comparison of integer expressions of different signedness: 'int' and 'long long unsigned int' [-Wsign-compare]
[Details] X:/BUILD/DOC_ROOT/ACE/ace/os_include/sys/os_socket.h:123:24: note: in expansion of macro 'ACE_CMSG_FIRSTHDR'
[Details] x:/BUILD/DOC_ROOT/ACE/ace/SOCK_Dgram.cpp:367:69: note: in expansion of macro 'ACE_CMSG_NXTHDR'
```

**Comment by mitza-oci** [comment]
> 
> 
> @mitza-oci It looks we have some new warnings triggered by this PR, see http://buildlogs.remedy.nl/win_mingw64_64_tao/2019_10_30_08_31_Brief.html#section_5
> 
> ```
> [Details] X:/BUILD/DOC_ROOT/ACE/ace/OS_NS_string.inl:312:20: warning: 'char* strncpy(char*, const char*, size_t)' specified bound 128 equals destination size [-Wstringop-truncation]
> [Details] x:/BUILD/DOC_ROOT/ACE/ace/OS_NS_sys_socket.cpp: In function 'void* ACE_6_5_6::ACE_OS::accept(ACE_HANDLE, sockaddr*, int*, const ACE_6_5_6::ACE_Accept_QoS_Params&)':
> [Details] x:/BUILD/DOC_ROOT/ACE/ace/OS_NS_sys_socket.cpp:25:90: warning: cast between incompatible function types from 'ACE_6_5_6::ACE_QOS_CONDITION_FUNC' {aka 'int (*)(iovec*, iovec*, ACE_6_5_6::ACE_QoS*, ACE_6_5_6::ACE_QoS*, iovec*, iovec*, unsigned int*, long unsigned int)'} to 'LPCONDITIONPROC' {aka 'int (*)(_WSABUF*, _WSABUF*, _QualityOfService*, _QualityOfService*, _WSABUF*, _WSABUF*, unsigned int*, long long unsigned int)'} [-Wcast-function-type]
> [Details] X:/BUILD/DOC_ROOT/ACE/ace/OS_NS_macros.h:28:34: note: in definition of macro 'ACE_SOCKCALL_RETURN'
> [Details] x:/BUILD/DOC_ROOT/ACE/ace/SOCK_Dgram.cpp: In member function 'ssize_t ACE_6_5_6::ACE_SOCK_Dgram::recv(iovec*, int, ACE_6_5_6::ACE_Addr&, int, ACE_6_5_6::ACE_INET_Addr*) const':
> [Details] X:/BUILD/DOC_ROOT/ACE/ace/os_include/sys/os_socket.h:120:30: warning: comparison of integer expressions of different signedness: 'int' and 'long long unsigned int' [-Wsign-compare]
> [Details] x:/BUILD/DOC_ROOT/ACE/ace/SOCK_Dgram.cpp:367:27: note: in expansion of macro 'ACE_CMSG_FIRSTHDR'
> [Details] X:/BUILD/DOC_ROOT/ACE/ace/os_include/sys/os_socket.h:120:30: warning: comparison of integer expressions of different signedness: 'int' and 'long long unsigned int' [-Wsign-compare]
> [Details] X:/BUILD/DOC_ROOT/ACE/ace/os_include/sys/os_socket.h:123:24: note: in expansion of macro 'ACE_CMSG_FIRSTHDR'
> [Details] x:/BUILD/DOC_ROOT/ACE/ace/SOCK_Dgram.cpp:367:69: note: in expansion of macro 'ACE_CMSG_NXTHDR'
> ```

Looks like 2 of these 3 are not related, but one warning (line 120 of os_socket.h) I can fix.

**Comment by jwillemsen** [comment]
@mitza-oci Please fix the new warnings so that we get down again to the previous number of warnings.

**Comment by jwillemsen** [comment]
The warnings are gone, thanks! Can you check the runtime test failures which I emailed directly @mitza-oci?

**Comment by jwillemsen** [comment]
New runtime failures of SOCK_Dgram_Test test
http://scoreboard.ociweb.com/doc_master_ant_linux-sb_gcc_d0o1/index.html
http://scoreboard.ociweb.com/doc_master_ant_linux-se_gcc_d0o1/index.html
http://scoreboard.ociweb.com/doc_master_hornet_win7-x64_vc14_i1d0o1w1/index.html

**Comment by mitza-oci** [comment]
> 
> 
> New runtime failures of SOCK_Dgram_Test test
> http://scoreboard.ociweb.com/doc_master_ant_linux-sb_gcc_d0o1/index.html
> http://scoreboard.ociweb.com/doc_master_ant_linux-se_gcc_d0o1/index.html
> http://scoreboard.ociweb.com/doc_master_hornet_win7-x64_vc14_i1d0o1w1/index.html

Fixed in #993

**Pull request merged**